aten/src/ATen/native/cpu/int4mm_kernel.cpp

```diff
@@ -0,0 +1,541 @@
+//
+// 2. avx2 or default packed format:
+//   Do 64-bit unpacking at a time.
+//
+//   weight: {a0, a1, a2, a3, ..., a14, a15}
+//   packed: {a1|a0, a3, a2, ..., a15|a14}
+//
+void weight_to_int4pack_kernel(
+    const Tensor& weight_packed,
+    const Tensor& weight) {
+
+  uint8_t* weight_packed_data = weight_packed.data_ptr<uint8_t>();
+  const int32_t* weight_data = weight.data_ptr<int32_t>();
+
+  int N = weight.size(0);
+  int K = weight.size(1);
+
+  // 64 for avx512 and 64 for avx2/non-vectorized
+  constexpr int BLOCK_N = vec::Vectorized<float>::size() * 4;
+  const int NB =  (N + BLOCK_N - 1) / BLOCK_N;
+
+  // parallel on NB blocks
+  at::parallel_for(0, NB, 0, [&](int begin, int end) {
+    for (const auto i : c10::irange(begin, end)) {
+      int nb_size = std::min(BLOCK_N, N - i * BLOCK_N);
+
+      const int32_t* src = weight_data + i * BLOCK_N * K;
+      uint8_t* dst = weight_packed_data + i * K * BLOCK_N / 2;
+      for (const auto k : c10::irange(K)) {
+#if defined(CPU_CAPABILITY_AVX512) && !defined(_MSC_VER)
+        if (nb_size == BLOCK_N) {
+          for (int d = 0; d < 16; d++) {
+            int32_t val0 = src[(d +  0) * K + k];
+            int32_t val1 = src[(d + 16) * K + k];
+            int32_t val2 = src[(d + 32) * K + k];
+            int32_t val3 = src[(d + 48) * K + k];
+
+            uint8_t packed02 = (((uint8_t)(val2) << 4)) | ((uint8_t)(val0));
+            uint8_t packed13 = (((uint8_t)(val3) << 4)) | ((uint8_t)(val1));
+
+            dst[k * 32 + d] = packed02;
+            dst[k * 32 + 16 + d] = packed13;
+          }
+        } else {
+          // for nb_size 16, 32, 48
+          for (int n = 0; n < nb_size; n += 2) {
+            int32_t val0 = src[n * K + k];
+            int32_t val1 = src[n * K + K + k];
+
+            uint8_t packed = (((uint8_t)(val1) << 4)) | ((uint8_t)(val0));
+            dst[k * nb_size / 2 + n / 2] = packed;
+          }
+        }
+#else
+        for (int n = 0; n < nb_size; n += 2) {
+          int32_t val0 = src[n * K + k];
+          int32_t val1 = src[n * K + K + k];
+
+          uint8_t packed = (((uint8_t)(val1) << 4)) | ((uint8_t)(val0));
+          dst[k * nb_size / 2 + n / 2] = packed;
+        }
+#endif
+      }
+    }
+  });
+}
+
+void int4pack_mm_kernel(
+    const Tensor& C,
+    const Tensor& A,
+    const Tensor& B,
+    int64_t qGroupSize,
+    const Tensor& qScaleAndZeros) {
+
+  const BFloat16* A_data = A.data_ptr<BFloat16>();
+  const uint8_t* B_data = B.data_ptr<uint8_t>();
+  BFloat16* C_data = C.data_ptr<BFloat16>();
+  const BFloat16* S_data = qScaleAndZeros.data_ptr<BFloat16>();
+
+  int M = A.size(0);
+  int N = B.size(0);
+  int K = A.size(1);
+
+  constexpr int BLOCK_M = 4;
+  // 64 for avx512 and 32 for avx2/non-vectorized
+  constexpr int BLOCK_N = vec::Vectorized<float>::size() * 4;
+  // 32, 64, 128, 256
+  const int BLOCK_K = qGroupSize;
+
+  const int MB = (M + BLOCK_M - 1) / BLOCK_M;
```

Comment:
Sure, will add this path later on.
The default benchmark from gpt fast has small prompt length so I planned to do it  later because this won't make much difference from the current benchmark.